File: cn/huohuas001/huhobotPenguin/spigot/events/GameChat.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  org.bukkit.Bukkit
 *  org.bukkit.command.CommandExecutor
 *  org.bukkit.command.PluginCommand
 *  org.bukkit.command.TabCompleter
 *  org.bukkit.event.EventHandler
 *  org.bukkit.event.EventPriority
 *  org.bukkit.event.Listener
 *  org.bukkit.event.player.AsyncPlayerChatEvent
 *  org.bukkit.event.player.PlayerJoinEvent
 *  org.bukkit.event.player.PlayerLoginEvent
 *  org.bukkit.event.player.PlayerLoginEvent$Result
 *  org.bukkit.event.player.PlayerQuitEvent
 *  org.bukkit.plugin.Plugin
 */
package cn.huohuas001.huhobotPenguin.spigot.events;

import cn.huohuas001.bot.QClient;
import cn.huohuas001.bot.events.commands.BindCommands;
import cn.huohuas001.huhobotPenguin.spigot.HuHoBotSpigot;
import cn.huohuas001.huhobotPenguin.spigot.commands.QqCommand;
import cn.huohuas001.huhobotPenguin.spigot.qqbind.AiChat;
import cn.huohuas001.huhobotPenguin.spigot.qqbind.QqBindManager;
import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.text.SimpleDateFormat;
import java.util.Date;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabCompleter;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerLoginEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.Plugin;

public final class GameChat
implements Listener {
    private static volatile boolean bindRegistered = false;
    private static int retryCount = 0;
    private static final int MAX_RETRIES = 8;

    public GameChat() {
        try {
            QqBindManager.getInstance();
        }
        catch (Throwable throwable) {
            QqBindManager.logQuiet("QqBindManager \u521d\u59cb\u5316\u5931\u8d25: " + GameChat.safeMessage(throwable));
        }
        GameChat.suppressForgeNetworkLogs();
        GameChat.redirectSdkLogs();
        GameChat.tryRegisterBind();
        Plugin plugin = Bukkit.getPluginManager().getPlugin("KERONGPenguin");
        if (plugin == null) {
            plugin = Bukkit.getPluginManager().getPlugin("KERONGPENGUIN");
        }
        if (plugin != null) {
            try {
                HuHoBotSpigot huHoBotSpigot;
                PluginCommand pluginCommand;
                if (plugin instanceof HuHoBotSpigot && (pluginCommand = (huHoBotSpigot = (HuHoBotSpigot)plugin).getCommand("qq")) != null) {
                    QqCommand qqCommand = new QqCommand(huHoBotSpigot);
                    pluginCommand.setExecutor((CommandExecutor)qqCommand);
                    pluginCommand.setTabCompleter((TabCompleter)qqCommand);
                    QqBindManager.logVerbose("[GameChat] /qq \u670d\u52a1\u5668\u547d\u4ee4\u5df2\u6ce8\u518c");
                }
            }
            catch (Throwable throwable) {
                QqBindManager.logQuiet("[GameChat] /qq \u6ce8\u518c\u5931\u8d25: " + GameChat.safeMessage(throwable));
            }
            Bukkit.getScheduler().runTaskLater(plugin, () -> {
                if (!bindRegistered) {
                    GameChat.tryRegisterBind();
                }
            }, 40L);
            Bukkit.getScheduler().runTaskLater(plugin, () -> {
                GameChat.checkEconomyForCheckin();
            }, 60L);
        }
    }

    private static synchronized void tryRegisterBind() {
        if (bindRegistered || retryCount >= 8) {
            return;
        }
        ++retryCount;
        try {
            QClient.INSTANCE.registerCommand(new BindCommands());
            bindRegistered = true;
            QqBindManager.logQuiet("/\u7ed1\u5b9a \u547d\u4ee4\u5df2\u6ce8\u518c (\u7b2c " + retryCount + " \u6b21\u5c1d\u8bd5)");
        }
        catch (Throwable throwable) {
            // empty catch block
        }
    }

    @EventHandler(ignoreCancelled=true)
    public final void onChat(AsyncPlayerChatEvent event) {
        String message = event.getMessage();
        String playerName = event.getPlayer().getName();
        QqBindManager mgr;
        try { mgr = QqBindManager.getInstance(); } catch (Throwable t) { mgr = null; }
        if (mgr != null && mgr.isServerAiEnabled() && mgr.isAiEnabled()) {
            String prefix = mgr.getServerAiPrefix();
            if (prefix != null && !prefix.isEmpty() && message.toLowerCase().startsWith(prefix + " ")) {
                String content = message.substring(prefix.length() + 1).trim();
                if (!content.isEmpty()) {
                    event.setCancelled(true);
                    final String fpName = playerName;
                    final String fContent = content;
                    final QqBindManager fMgr = mgr;
                    final String fPlayerName = fpName;
                    final String fContent2 = fContent;
                    final QqBindManager fMgr2 = fMgr;
                    Bukkit.getScheduler().runTaskAsynchronously(Bukkit.getPluginManager().getPlugin("KERONGPenguin"), new Runnable() {
                        @Override
                        public void run() {
                            try {
                                final String aiReply = AiChat.chat(fContent2, fMgr2, "server", fPlayerName);
                                final String aiOutPrefix = fMgr2.getServerAiOutputPrefix();
                                Bukkit.getScheduler().runTask(Bukkit.getPluginManager().getPlugin("KERONGPenguin"), new Runnable() {
                                    @Override
                                    public void run() {
                                        Bukkit.broadcastMessage(aiOutPrefix + " " + aiReply);
                                    }
                                });
                        } catch (final Throwable t2) {
                            QqBindManager.logQuiet("[AI\u5bf9\u8bdd] \u670d\u52a1\u5668\u8c03\u7528\u5931\u8d25: " + t2.getMessage());
                            final String errPrefix = fMgr2.getServerAiOutputPrefix();
                            final String errMsg = t2.getMessage();
                            Bukkit.getScheduler().runTask(Bukkit.getPluginManager().getPlugin("KERONGPenguin"), new Runnable() {
                                @Override
                                public void run() {
                                    Bukkit.broadcastMessage(errPrefix + " AI \u5bf9\u8bdd\u5931\u8d25: " + errMsg);
                                }
                            });
                        }
                    }
                    });
                    return;
                }
            }
        }
        QClient.INSTANCE.broadcastGameMessage(playerName, message);
    }

    @EventHandler
    public final void onPlayerJoin(PlayerJoinEvent playerJoinEvent) {
        final String playerName = playerJoinEvent.getPlayer().getName();
        QClient.INSTANCE.broadcastPlayerJoin(playerName);
        Plugin plugin = Bukkit.getPluginManager().getPlugin("KERONGPenguin");
        if (plugin == null) {
            return;
        }
        Bukkit.getScheduler().runTaskAsynchronously(plugin, new Runnable() {
            @Override
            public void run() {
                try {
                    QqBindManager mgr = QqBindManager.getInstance();
                    String quuid = mgr.getQuuid(playerName);
                    if (quuid == null || quuid.isEmpty()) {
                        return;
                    }
                    final double pending = mgr.takePendingCoins(quuid);
                    if (pending > 0) {
                        Bukkit.getScheduler().runTask(Bukkit.getPluginManager().getPlugin("KERONGPenguin"), new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    org.bukkit.entity.Player player = Bukkit.getPlayerExact(playerName);
                                    if (player != null && player.isOnline()) {
                                        boolean ok = depositToPlayer(playerName, pending);
                                        if (ok) {
                                            player.sendMessage("\u00a7a[\u7b7e\u5230\u5956\u52b1] \u9886\u53d6\u7b7e\u5230\u5956\u52b1 " + formatMoney(pending) + " \u91d1\u5e01");
                                        } else {
                                            QqBindManager.logQuiet("[\u7b7e\u5230\u5956\u52b1] \u7ed9\u4e88\u73a9\u5bb6 " + playerName + " \u91d1\u5e01\u5931\u8d25\uff08Vault\uff09");
                                        }
                                    }
                                } catch (Throwable t) {
                                    QqBindManager.logQuiet("[\u7b7e\u5230\u5956\u52b1] \u73a9\u5bb6\u4e0a\u7ebf\u9886\u53d6\u5f02\u5e38: " + GameChat.safeMessage(t));
                                }
                            }
                        });
                    }
                } catch (Throwable t) {
                    // empty catch block
                }
            }
        });
    }

    public static boolean depositToPlayer(String playerName, double amount) {
        try {
            Plugin vault = Bukkit.getPluginManager().getPlugin("Vault");
            if (vault == null) {
                return false;
            }
            org.bukkit.OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(playerName);
            Class<?> economyClass = Class.forName("net.milkbowl.vault.economy.Economy");
            Class<?> registeredServiceProviderClass = Class.forName("org.bukkit.plugin.RegisteredServiceProvider");
            Method getServicesManager = Bukkit.getServer().getClass().getMethod("getServicesManager", new Class[0]);
            Object servicesManager = getServicesManager.invoke(Bukkit.getServer(), new Object[0]);
            Method getRegistration = servicesManager.getClass().getMethod("getRegistration", Class.class);
            Object rsp = getRegistration.invoke(servicesManager, economyClass);
            if (rsp == null) {
                return false;
            }
            Method getProvider = registeredServiceProviderClass.getMethod("getProvider", new Class[0]);
            Object economy = getProvider.invoke(rsp, new Object[0]);
            if (economy == null) {
                return false;
            }
            Method depositPlayer = economyClass.getMethod("depositPlayer", org.bukkit.OfflinePlayer.class, Double.TYPE);
            Object result = depositPlayer.invoke(economy, offlinePlayer, amount);
            if (result == null) {
                return false;
            }
            try {
                Method transactionSuccess = result.getClass().getMethod("transactionSuccess", new Class[0]);
                Object success = transactionSuccess.invoke(result, new Object[0]);
                return Boolean.TRUE.equals(success);
            } catch (Throwable throwable) {
                return true;
            }
        } catch (Throwable throwable) {
            return false;
        }
    }

    public static String formatMoney(double amount) {
        if (amount == (long)amount) {
            return String.valueOf((long)amount);
        }
        return String.valueOf(amount);
    }

    private static void checkEconomyForCheckin() {
        try {
            QqBindManager mgr = QqBindManager.getInstance();
            if (!mgr.isCheckinEnabled()) {
                return;
            }
            Plugin vault = Bukkit.getPluginManager().getPlugin("Vault");
            if (vault != null && vault.isEnabled()) {
                boolean economyReady = false;
                try {
                    Class<?> economyClass = Class.forName("net.milkbowl.vault.economy.Economy");
                    Method getServicesManager = Bukkit.getServer().getClass().getMethod("getServicesManager", new Class[0]);
                    Object servicesManager = getServicesManager.invoke(Bukkit.getServer(), new Object[0]);
                    Method getRegistration = servicesManager.getClass().getMethod("getRegistration", Class.class);
                    Object rsp = getRegistration.invoke(servicesManager, economyClass);
                    if (rsp != null) {
                        Class<?> registeredServiceProviderClass = Class.forName("org.bukkit.plugin.RegisteredServiceProvider");
                        Method getProvider = registeredServiceProviderClass.getMethod("getProvider", new Class[0]);
                        Object economy = getProvider.invoke(rsp, new Object[0]);
                        if (economy != null) {
                            economyReady = true;
                        }
                    }
                } catch (Throwable t) {
                    // empty catch block
                }
                if (economyReady) {
                    Bukkit.getConsoleSender().sendMessage("\u00a7a[KERONGPenguin] \u7b7e\u5230\u529f\u80fd\u5df2\u5f00\u542f\uff0c\u5df2\u68c0\u6d4b\u5230\u7ecf\u6d4e\u63d2\u4ef6 Vault\uff08Economy \u5c31\u7eea\uff09\u3002");
                } else {
                    Bukkit.getConsoleSender().sendMessage("\u00a7c[KERONGPenguin] \u7b7e\u5230\u529f\u80fd\u5df2\u5f00\u542f\uff0c\u4f46\u672a\u68c0\u6d4b\u5230\u7ecf\u6d4e\u524d\u7f6e\uff08Economy \u672a\u5c31\u7eea\uff09\u3002\u8bf7\u5b89\u88c5 Vault \u53ca\u4efb\u4e00\u7ecf\u6d4e\u63d2\u4ef6\uff08\u5982 EssentialsX\uff09\u3002");
                }
            } else {
                Bukkit.getConsoleSender().sendMessage("\u00a7c[KERONGPenguin] \u7b7e\u5230\u529f\u80fd\u5df2\u5f00\u542f\uff0c\u4f46\u672a\u68c0\u6d4b\u5230 Vault \u7ecf\u6d4e\u63d2\u4ef6\u3002\u8bf7\u5b89\u88c5 Vault \u53ca\u4efb\u4e00\u7ecf\u6d4e\u63d2\u4ef6\uff08\u5982 EssentialsX\uff09\u3002");
            }
        } catch (Throwable t) {
            QqBindManager.logQuiet("[GameChat] \u7b7e\u5230\u7ecf\u6d4e\u68c0\u6d4b\u5931\u8d25: " + GameChat.safeMessage(t));
        }
    }

    @EventHandler
    public final void onPlayerQuit(PlayerQuitEvent playerQuitEvent) {
        QClient.INSTANCE.broadcastPlayerQuit(playerQuitEvent.getPlayer().getName());
    }

    @EventHandler(priority=EventPriority.LOW)
    public final void onPlayerLogin(PlayerLoginEvent playerLoginEvent) {
        QqBindManager qqBindManager;
        if (!bindRegistered) {
            GameChat.tryRegisterBind();
        }
        try {
            qqBindManager = QqBindManager.getInstance();
        }
        catch (Throwable throwable) {
            return;
        }
        if (!qqBindManager.isEnabled()) {
            return;
        }
        String string = playerLoginEvent.getPlayer().getName();
        if (qqBindManager.isSkipped(string)) {
            return;
        }
        qqBindManager.getOrCreateQuuid(string);
        if (qqBindManager.isBound(string)) {
            return;
        }
        String string2 = qqBindManager.generateCode(string);
        playerLoginEvent.disallow(PlayerLoginEvent.Result.KICK_OTHER, qqBindManager.formatKickMessage(string2, string));
    }

    private static String safeMessage(Throwable throwable) {
        String string = throwable.getMessage();
        return string == null || string.isEmpty() ? throwable.toString() : string;
    }

    private static void suppressForgeNetworkLogs() {
        try {
            String[] stringArray = new String[]{"net.minecraftforge.network", "net.minecraftforge.network.NetworkEvent", "net.minecraftforge.network.HandshakeMessages", "net.minecraftforge.fml", "net.minecraftforge.fml.network", "fml", "FMLHandshakeHandler", "net.minecraft.network", "net.minecraftforge.fml.network.FMLNetworkConstants"};
            Class<?> clazz = null;
            Class<?> clazz2 = null;
            Class<?> clazz3 = null;
            try {
                clazz = Class.forName("org.apache.logging.log4j.LogManager");
                clazz2 = Class.forName("org.apache.logging.log4j.Level");
                clazz3 = Class.forName("org.apache.logging.log4j.core.config.Configurator");
            }
            catch (Throwable throwable) {
                // empty catch block
            }
            if (clazz != null && clazz2 != null && clazz3 != null) {
                Method method = clazz3.getMethod("setLevel", String.class, clazz2);
                Field field = clazz2.getField("WARN");
                Object object = field.get(null);
                for (String string : stringArray) {
                    try {
                        method.invoke(null, string, object);
                    }
                    catch (Throwable throwable) {
                        // empty catch block
                    }
                }
                QqBindManager.logQuiet("[GameChat] \u5df2\u6291\u5236 Forge \u7f51\u7edc\u65e5\u5fd7");
            }
        }
        catch (Throwable throwable) {
            // empty catch block
        }
    }

    private static void redirectSdkLogs() {
        try {
            Class<?> clazz = Class.forName("io.github.kloping.qqbot.utils.LoggerImpl$LogSink");
            Class<?> clazz2 = Class.forName("io.github.kloping.qqbot.utils.LoggerImpl");
            Method method2 = clazz2.getMethod("setLogSink", clazz);
            Object object2 = Proxy.newProxyInstance(clazz.getClassLoader(), new Class[]{clazz}, (object, method, objectArray) -> {
                block8: {
                    if (method.getName().equals("log") && objectArray != null && objectArray.length >= 1) {
                        String string = String.valueOf(objectArray[0]);
                        try {
                            File file = GameChat.getBotLogFile();
                            if (file == null) break block8;
                            try (PrintWriter printWriter = new PrintWriter(new FileWriter(file, true));){
                                String string2 = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
                                printWriter.println("[" + string2 + "] [Bot] " + string);
                            }
                        }
                        catch (Throwable throwable) {
                            // empty catch block
                        }
                    }
                }
                return null;
            });
            method2.invoke(null, object2);
            QqBindManager.logQuiet("[GameChat] \u5df2\u91cd\u5b9a\u5411 Bot SDK \u65e5\u5fd7\u5230\u6587\u4ef6");
        }
        catch (Throwable throwable) {
            // empty catch block
        }
    }

    private static File getBotLogFile() {
        try {
            QqBindManager qqBindManager = QqBindManager.getInstance();
            Method method = QqBindManager.class.getDeclaredMethod("getLogFile", new Class[0]);
            method.setAccessible(true);
            return (File)method.invoke(null, new Object[0]);
        }
        catch (Throwable throwable) {
            return null;
        }
    }
}

File: server-Spigot/src/main/kotlin/cn/huohuas001/huhobotPenguin/spigot/events/GameChat.kt
package cn.huohuas001.huhobotPenguin.spigot.events

import cn.huohuas001.bot.QClient
import cn.huohuas001.bot.events.commands.BindCommands
import cn.huohuas001.huhobotPenguin.spigot.commands.QqCommand
import cn.huohuas001.huhobotPenguin.spigot.qqbind.AiChat
import cn.huohuas001.huhobotPenguin.spigot.qqbind.QqBindManager
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerLoginEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.Plugin
import java.text.SimpleDateFormat
import java.util.Date

/**
 * 游戏事件监听：
 * - 聊天转发到 QQ 群（含服务器内 AI 对话前缀）；
 * - 玩家进入/退出播报（进入时补发离线签到奖励）；
 * - 未绑定 QQ 的玩家登录时拒入并发绑定码；
 * - 抑制 Forge 网络日志刷屏、重定向 Bot SDK 日志到文件。
 */
class GameChat : Listener {
    internal companion object {
        @Volatile
        private var bindRegistered = false
        private var retryCount = 0
        private const val MAX_RETRIES = 8

        /** Vault 入账（反射，避免编译期依赖）。 */
        @JvmStatic
        fun depositToPlayer(playerName: String, amount: Double): Boolean = try {
            val vault = Bukkit.getPluginManager().getPlugin("Vault") ?: return false
            val offlinePlayer = Bukkit.getOfflinePlayer(playerName)
            val economyClass = Class.forName("net.milkbowl.vault.economy.Economy")
            val registeredServiceProviderClass = Class.forName("org.bukkit.plugin.RegisteredServiceProvider")
            val servicesManager = Bukkit.getServer().javaClass.getMethod("getServicesManager").invoke(Bukkit.getServer())
            val rsp = servicesManager.javaClass.getMethod("getRegistration", Class::class.java)
                .invoke(servicesManager, economyClass) ?: return false
            val economy = registeredServiceProviderClass.getMethod("getProvider").invoke(rsp) ?: return false
            val result = economyClass
                .getMethod("depositPlayer", org.bukkit.OfflinePlayer::class.java, Double::class.javaPrimitiveType)
                .invoke(economy, offlinePlayer, amount) ?: return false
            try {
                val success = result.javaClass.getMethod("transactionSuccess").invoke(result)
                java.lang.Boolean.TRUE == success
            } catch (_: Throwable) {
                true
            }
        } catch (_: Throwable) {
            false
        }

        /** 金额显示（整数去掉小数点）。 */
        @JvmStatic
        fun formatMoney(amount: Double): String =
            if (amount == amount.toLong().toDouble()) amount.toLong().toString() else amount.toString()
    }

    init {
        try {
            QqBindManager.getInstance()
        } catch (t: Throwable) {
            QqBindManager.logQuiet("QqBindManager 初始化失败: ${safeMessage(t)}")
        }
        suppressForgeNetworkLogs()
        redirectSdkLogs()
        tryRegisterBind()

        var plugin: Plugin? = Bukkit.getPluginManager().getPlugin("KERONGPenguin")
        if (plugin == null) {
            plugin = Bukkit.getPluginManager().getPlugin("KERONGPENGUIN")
        }
        if (plugin != null) {
            try {
                if (plugin is cn.huohuas001.huhobotPenguin.spigot.HuHoBotSpigot) {
                    plugin.getCommand("qq")?.let { command ->
                        val qqCommand = QqCommand(plugin)
                        command.setExecutor(qqCommand)
                        command.setTabCompleter(qqCommand)
                        QqBindManager.logVerbose("[GameChat] /qq 服务器命令已注册")
                    }
                }
            } catch (t: Throwable) {
                QqBindManager.logQuiet("[GameChat] /qq 注册失败: ${safeMessage(t)}")
            }
            // 延迟补注册 /绑定 命令（等待 QQ 客户端启动）
            Bukkit.getScheduler().runTaskLater(plugin, Runnable {
                if (!bindRegistered) tryRegisterBind()
            }, 40L)
            // 延迟检测经济插件（签到功能依赖）
            Bukkit.getScheduler().runTaskLater(plugin, Runnable { checkEconomyForCheckin() }, 60L)
        }
    }

    /** 尝试向 QQ 客户端注册 /绑定 命令。 */
    @Synchronized
    private fun tryRegisterBind() {
        if (bindRegistered || retryCount >= MAX_RETRIES) return
        retryCount++
        try {
            QClient.registerCommand(BindCommands())
            bindRegistered = true
            QqBindManager.logQuiet("/绑定 命令已注册 (第 $retryCount 次尝试)")
        } catch (_: Throwable) {
        }
    }

    /** 群聊转发 + 服务器内 AI 对话。 */
    @EventHandler(ignoreCancelled = true)
    fun onChat(event: AsyncPlayerChatEvent) {
        val message = event.message
        val playerName = event.player.name
        val manager = try {
            QqBindManager.getInstance()
        } catch (_: Throwable) {
            null
        }
        // 服务器内 AI 对话（前缀触发）
        if (manager != null && manager.isServerAiEnabled && manager.isAiEnabled) {
            val prefix = manager.serverAiPrefix()
            if (prefix.isNotEmpty() && message.lowercase().startsWith("$prefix ")) {
                val content = message.substring(prefix.length + 1).trim()
                if (content.isNotEmpty()) {
                    event.isCancelled = true
                    val bukkitPlugin = Bukkit.getPluginManager().getPlugin("KERONGPenguin")
                        ?: return
                    Bukkit.getScheduler().runTaskAsynchronously(bukkitPlugin, Runnable {
                        try {
                            val reply = AiChat.chat(content, manager, "server", playerName)
                            val outputPrefix = manager.serverAiOutputPrefix()
                            Bukkit.getScheduler().runTask(bukkitPlugin, Runnable {
                                Bukkit.broadcastMessage("$outputPrefix $reply")
                            })
                        } catch (t: Throwable) {
                            QqBindManager.logQuiet("[AI对话] 服务器调用失败: ${t.message}")
                            val outputPrefix = manager.serverAiOutputPrefix()
                            Bukkit.getScheduler().runTask(bukkitPlugin, Runnable {
                                Bukkit.broadcastMessage("$outputPrefix AI 对话失败: ${t.message}")
                            })
                        }
                    })
                    return
                }
            }
        }
        QClient.broadcastGameMessage(playerName, message)
    }

    /** 玩家进入：播报 + 补发离线签到奖励。 */
    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val playerName = event.player.name
        val playerUuid = event.player.uniqueId.toString()
        QClient.broadcastPlayerJoin(playerName)
        val plugin = Bukkit.getPluginManager().getPlugin("KERONGPenguin") ?: return

        Bukkit.getScheduler().runTaskAsynchronously(plugin, Runnable {
            try {
                val manager = QqBindManager.getInstance()
                var quuid = manager.getQuuidByUuid(playerName, playerUuid)
                if (quuid.isNullOrEmpty()) {
                    // 兼容旧版：按纯玩家名查找
                    quuid = manager.getQuuid(playerName)
                }
                if (quuid.isNullOrEmpty()) return@Runnable

                // 补发离线签到奖励
                val pending = manager.takePendingCoins(quuid)
                if (pending > 0) {
                    Bukkit.getScheduler().runTask(plugin, Runnable {
                        try {
                            val player = Bukkit.getPlayerExact(playerName)
                            if (player != null && player.isOnline) {
                                val ok = depositToPlayer(playerName, pending)
                                if (ok) {
                                    player.sendMessage("§a[签到奖励] 领取签到奖励 ${formatMoney(pending)} 金币")
                                } else {
                                    QqBindManager.logQuiet("[签到奖励] 给予玩家 $playerName 金币失败（Vault）")
                                }
                            }
                        } catch (t: Throwable) {
                            QqBindManager.logQuiet("[签到奖励] 玩家上线领取异常: ${safeMessage(t)}")
                        }
                    })
                }
            } catch (_: Throwable) {
            }
        })
    }

    /** 玩家退出：播报。 */
    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        QClient.broadcastPlayerQuit(event.player.name)
    }

    /** 未绑定 QQ 的玩家登录拒入（UUID 绑定机制）。 */
    @EventHandler(priority = EventPriority.LOW)
    fun onPlayerLogin(event: PlayerLoginEvent) {
        if (!bindRegistered) {
            tryRegisterBind()
        }
        val manager = try {
            QqBindManager.getInstance()
        } catch (_: Throwable) {
            return
        }
        if (!manager.isEnabled) return
        val playerName = event.player.name
        val playerUuid = event.player.uniqueId.toString()
        if (manager.isSkipped(playerName)) return

        // 确保绑定记录存在（UUID 机制，含旧数据迁移）
        manager.getOrCreateQuuidByUuid(playerName, playerUuid)
        if (manager.isBoundByUuid(playerName, playerUuid)) return

        // 未绑定：生成绑定码并拒入
        val code = manager.generateCode(playerName, playerUuid)
        event.disallow(PlayerLoginEvent.Result.KICK_OTHER, manager.formatKickMessage(code, playerName))
    }

    /** 启动时检测经济插件可用性（签到功能提示）。 */
    private fun checkEconomyForCheckin() {
        try {
            val manager = QqBindManager.getInstance()
            if (!manager.isCheckinEnabled) return
            val vault = Bukkit.getPluginManager().getPlugin("Vault")
            if (vault != null && vault.isEnabled) {
                val economyReady = try {
                    val economyClass = Class.forName("net.milkbowl.vault.economy.Economy")
                    val servicesManager = Bukkit.getServer().javaClass.getMethod("getServicesManager").invoke(Bukkit.getServer())
                    val rsp = servicesManager.javaClass.getMethod("getRegistration", Class::class.java)
                        .invoke(servicesManager, economyClass)
                    if (rsp != null) {
                        val registeredServiceProviderClass = Class.forName("org.bukkit.plugin.RegisteredServiceProvider")
                        registeredServiceProviderClass.getMethod("getProvider").invoke(rsp) != null
                    } else false
                } catch (_: Throwable) {
                    false
                }
                if (economyReady) {
                    Bukkit.getConsoleSender().sendMessage("§a[KERONGPenguin] 签到功能已开启，已检测到经济插件 Vault（Economy 就绪）。")
                } else {
                    Bukkit.getConsoleSender().sendMessage("§c[KERONGPenguin] 签到功能已开启，但未检测到经济前置（Economy 未就绪）。请安装 Vault 及任一经济插件（如 EssentialsX）。")
                }
            } else {
                Bukkit.getConsoleSender().sendMessage("§c[KERONGPenguin] 签到功能已开启，但未检测到 Vault 经济插件。请安装 Vault 及任一经济插件（如 EssentialsX）。")
            }
        } catch (t: Throwable) {
            QqBindManager.logQuiet("[GameChat] 签到经济检测失败: ${safeMessage(t)}")
        }
    }

    private fun safeMessage(throwable: Throwable): String =
        throwable.message?.takeIf { it.isNotEmpty() } ?: throwable.toString()

    /** 抑制 Forge 网络握手日志刷屏（混合端）。 */
    private fun suppressForgeNetworkLogs() {
        try {
            val loggers = listOf(
                "net.minecraftforge.network", "net.minecraftforge.network.NetworkEvent",
                "net.minecraftforge.network.HandshakeMessages", "net.minecraftforge.fml",
                "net.minecraftforge.fml.network", "fml", "FMLHandshakeHandler",
                "net.minecraft.network", "net.minecraftforge.fml.network.FMLNetworkConstants",
            )
            val logManager = try {
                Class.forName("org.apache.logging.log4j.LogManager")
            } catch (_: Throwable) {
                null
            } ?: return
            val levelClass = try {
                Class.forName("org.apache.logging.log4j.Level")
            } catch (_: Throwable) {
                null
            } ?: return
            val configurator = try {
                Class.forName("org.apache.logging.log4j.core.config.Configurator")
            } catch (_: Throwable) {
                null
            } ?: return
            val setLevel = configurator.getMethod("setLevel", String::class.java, levelClass)
            val warnLevel = levelClass.getField("WARN").get(null)
            for (name in loggers) {
                try {
                    setLevel.invoke(null, name, warnLevel)
                } catch (_: Throwable) {
                }
            }
            QqBindManager.logQuiet("[GameChat] 已抑制 Forge 网络日志")
        } catch (_: Throwable) {
        }
    }

    /** 将 Bot SDK 日志重定向到文件（不刷服务器控制台）。 */
    private fun redirectSdkLogs() {
        try {
            val sinkClass = Class.forName("io.github.kloping.qqbot.utils.LoggerImpl\$LogSink")
            val loggerClass = Class.forName("io.github.kloping.qqbot.utils.LoggerImpl")
            val setLogSink = loggerClass.getMethod("setLogSink", sinkClass)
            val sink = java.lang.reflect.Proxy.newProxyInstance(
                sinkClass.classLoader, arrayOf(sinkClass),
            ) { _, method, args ->
                if (method.name == "log" && args != null && args.isNotEmpty()) {
                    val message = args[0]?.toString() ?: ""
                    try {
                        val file = QqBindManager.currentLogFile()
                        if (file != null) {
                            java.io.PrintWriter(java.io.FileWriter(file, true)).use { writer ->
                                val timestamp = SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(Date())
                                writer.println("[$timestamp] [Bot] $message")
                            }
                        }
                    } catch (_: Throwable) {
                    }
                }
                null
            }
            setLogSink.invoke(null, sink)
            QqBindManager.logQuiet("[GameChat] 已重定向 Bot SDK 日志到文件")
        } catch (_: Throwable) {
        }
    }
}
